platform-sdk/docs/proposals/consensus-layer/Consensus-Layer.md

```diff
@@ -0,0 +1,438 @@
+module so the offending node may be disciplined. Note that the node to be disciplined will be the node that sent this
+bad event to us, not the origin node. This information (which node sent the event) must be captured by Gossip and
+passed to Event Intake as part of the event metadata.
+
+#### Topological Ordering
+
+Events are buffered if necessary to ensure that each parent event has been emitted from Event Intake before any child
+events. A simple map (the same used for deduplication) can be used here. Given some event, for each parent, look up the
+parent by its hash. If each parent is found in the map, then emit the event. Otherwise, remember the event so when the
+missing parent is received, the child may be emitted. The current implementation uses what is known as the "orphan
+buffer" for this purpose.
+
+Since Event Intake will also maintain some buffers, it needs to know about the progression of the hashgraph,
+so it can evict old events. In this case, the "orphan buffer" holds events until either the parent events have
+arrived, or the events expired due to the advancement of the "non-ancient event window" and the event is
+dropped from the buffer. This document does not prescribe the existence of the orphan buffer or the method by which
+events are sorted and emitted in topological order, but it does describe a method by which old events can be dropped.
+
+#### Assigning Generations
+
+Several algorithms, including the hashgraph consensus algorithm and the tipset algorithm, require events to be part of
+a "generation". We use what is known as "local generation". Each node, when it starts, assigns "1" as the generation
+of the first event it encounters per event creator. So the first event from Alice is generation 1, as is the first event
+from Bob, but the second event from Alice is in generation 2. This resets each time the consensus module is restarted
+(node restart, upgrade, reconnect, etc.). The absolute numbers for the generations are not important, but their
+relative values are critical. Generations always increase by 1. Different nodes may assign different generations to
+the same events. The generation will be part of the event metadata, **not** part of `EventCore` (that is, it is not
+in the part of the event that is gossiped).
+
+Event Intake is responsible for assigning generations to events.
+
+#### Fork Detection
+
+The Event Intake module inspects events to determine whether any given event creator is "forking" the hashgraph. A
+"fork" happens when two or more different events from the same creator have the same "self-event" parent. Any node
+that forks (known affectionately as a "Dirty Rotten Forker") will be reported to the Sheriff. Forking is a sign of
+either a dishonest node, or a seriously broken node. In either case, it may be subject to "shunning", and will be
+reported to the Execution layer for further observation and, if required, action (such as canceling rewards for
+stakers to that node).
```

Comment:
Although the platform doesn't currently do it, in order to be secure against branching attacks, there will need to be branch detection+handling post-consensus in addition to pre-consensus. Pre-consensus detection happens quickly and allows an attacking party to be shunned, but it is not provable. Post-consensus detection may take a few seconds (especially in the presence of a branching attack), but all honest nodes will come to the same conclusion, and can take network wide action to eject the attacker.